### PR TITLE
Make fields for expression public & Add finer literal parsing

### PR DESCRIPTION
**Pull request by krishnangovindraj**
## Usage and product changes
Make fields for expression public + Add finer literal parsing

## Implementation
We now recurse to the leaves of the grammar when parsing literals. The literals hold the parts instead of a single string.

### CODE REVIEW

File: rust/parser/expression.rs [hunk 1]
```diff
@@ -97,7 +97,7 @@ fn visit_list_index(node: Node<'_>) -> Expression {
 
 pub(super) fn visit_expression_struct(node: Node<'_>) -> Literal {
     debug_assert_eq!(node.as_rule(), Rule::expression_struct);
-    Literal::new(node.span(), None, node.as_str().to_owned()) // TODO parse properly
+    Literal::new(node.span(), todo!()) // TODO parse properly
```
Comment (krishnangovindraj) [inline rust/parser/expression.rs:100]: What are the struct expressions? The declaration?

**Review by flyingsilverfin — APPROVED** [review]

File: rust/parser/mod.rs [hunk 2]
```diff
@@ -305,15 +309,111 @@ fn visit_value_type_primitive(node: Node<'_>) -> BuiltinValueType {
 
 fn visit_value_literal(node: Node<'_>) -> Literal {
```
Comment (dmitrii-ubskii) [inline rust/parser/mod.rs:310]: At this point literal visitors warrant extraction into their own module

File: rust/parser/mod.rs [hunk 3]
```diff
@@ -305,15 +309,111 @@ fn visit_value_type_primitive(node: Node<'_>) -> BuiltinValueType {
...
 }
 
-fn visit_quoted_string_literal(node: Node<'_>) -> Literal {
-    debug_assert_eq!(node.as_rule(), Rule::quoted_string_literal);
-    Literal::new(node.span(), Some(Tag::String), node.as_str().to_owned())
+fn visit_sign(node: Node<'_>) -> Sign {
+    debug_assert_eq!(node.as_rule(), Rule::sign);
+    match node.as_rule() {
+        Rule::PLUS => Sign::Plus,
+        Rule::MIN => Sign::Minus,
```
Comment (dmitrii-ubskii) [inline rust/parser/mod.rs:333]: This is wrong. Run `cargo test -- autogenerated`, it is meant to catch typos like these

File: rust/parser/expression.rs (hunk 1 above)
Comment (dmitrii-ubskii) [inline rust/parser/expression.rs:100]: Is there a non-panicking stopgap you can put here for now?

File: rust/query/pipeline/stage/modifier.rs [hunk 4]
```diff
@@ -13,7 +13,7 @@ use crate::{
     },
     pretty::Pretty,
     util::write_joined,
-    value::Literal,
+    value::{IntegerLiteral, Literal},
```
Comment (dmitrii-ubskii) [inline rust/query/pipeline/stage/modifier.rs:16]: Is `Literal` even used here anymore?

File: rust/value.rs [hunk 5]
```diff
@@ -4,66 +4,138 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
  */
 
-use std::fmt;
+use std::{
+    fmt,
+    fmt::{Debug, Formatter},
```
Comment (dmitrii-ubskii) [inline rust/value.rs:9]: Nix

File: rust/value.rs [hunk 6]
```diff
@@ -4,66 +4,138 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
  */
 
-use std::fmt;
+use std::{
+    fmt,
```
Comment (dmitrii-ubskii) [inline rust/value.rs:8]: `fmt::self`

File: rust/value.rs [hunk 7]
```diff
@@ -4,66 +4,138 @@
...
+    fmt::{Debug, Formatter},
+};
+use std::fmt::Write;
+use std::path::Display;
```
Comment (dmitrii-ubskii) [inline rust/value.rs:12]: Format

File: rust/value.rs [hunk 8]
```diff
@@ -76,8 +148,37 @@ impl Spanned for Literal {
 impl Pretty for Literal {}
 
 impl fmt::Display for Literal {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str(&self.inner)
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        self.inner.fmt(f)
```
Comment (dmitrii-ubskii) [inline rust/value.rs:152]: This is ambiguous between `Display` and `Debug` (probably where the import came from), should be `fmt::Display::fmt(self.inner, f)`

File: rust/value.rs [hunk 9]
```diff
@@ -76,8 +148,37 @@ impl Spanned for Literal {
...
+        match self {
+            Sign::Plus => f.write_char('+'),
+            Sign::Minus => f.write_char('-'),
+        }
+    }
+}
+
+impl fmt::Display for SignedDecimalLiteral {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        std::fmt::Display::fmt(&self.sign, f)?;
```
Comment (dmitrii-ubskii) [inline rust/value.rs:179]:
```suggestion
        fmt::Display::fmt(&self.sign, f)?;
```

**Review by dmitrii-ubskii — CHANGES_REQUESTED** [review]

File: rust/parser/mod.rs (hunk 3 above)
Comment (krishnangovindraj) [reply rust/parser/mod.rs]: Ah, this didn't run through the IDE. I guess that's why I didn't catch them.

**Review by krishnangovindraj — COMMENTED** [review]
replies

File: rust/value.rs [hunk 10]
```diff
@@ -76,57 +143,225 @@ impl Spanned for Literal {
...
+}
+
+impl fmt::Display for StructLiteral {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str(self.inner.as_str())
+    }
+}
+
+impl StringLiteral {
+    pub fn parse_to_string(escaped_string: &str) -> Result<String> {
```
Comment (krishnangovindraj) [inline rust/value.rs:318]: I've moved this to an impl of StringLiteral, which doesn't feel quite right.

File: rust/parser/annotation.rs [hunk 11]
```diff
@@ -4,15 +4,13 @@
...
-    visit_identifier, visit_integer_literal, visit_quoted_string_literal, visit_value_literal, IntoChildNodes, Node,
-    Rule, RuleMatcher,
-};
+use super::{visit_identifier, IntoChildNodes, Node, Rule, RuleMatcher};
 use crate::{
     annotation::{
         Abstract, Annotation, Cardinality, Cascade, Distinct, Independent, Key, Range, Regex, Subkey, Unique, Values,
     },
     common::{error::TypeQLError, Spanned},
+    parser::literal::{visit_integer_literal, visit_quoted_string_literal, visit_value_literal},
```
Comment (dmitrii-ubskii) [inline rust/parser/annotation.rs:13]: `crate::parser` is `super`, should be merged with the `use` above.

File: rust/parser/expression.rs [hunk 12]
```diff
@@ -6,13 +6,14 @@
...
 
-use super::{visit_identifier, visit_value_literal, visit_var, IntoChildNodes, Node, Rule, RuleMatcher};
+use super::{visit_identifier, visit_var, IntoChildNodes, Node, Rule, RuleMatcher};
 use crate::{
     common::{error::TypeQLError, token, Spanned},
     expression::{
         BuiltinFunctionName, Expression, FunctionCall, FunctionName, List, ListIndex, ListIndexRange, Operation, Paren,
     },
-    value::Literal,
+    parser::literal::visit_value_literal,
```
Comment (dmitrii-ubskii) [inline rust/parser/expression.rs:15]: As above

File: rust/parser/literal.rs [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        Rule::date_literal => ValueLiteral::Date(visit_date_literal(child)),
+
+        _ => unreachable!("{}", TypeQLError::IllegalGrammar { input: child.to_string() }),
+    };
+    Literal::new(span, value_literal)
+}
+
+fn visit_sign(node: Node<'_>) -> Sign {
+    debug_assert_eq!(node.as_rule(), Rule::sign);
+    match node.as_str() {
```
Comment (dmitrii-ubskii) [inline rust/parser/literal.rs:37]: We should be matching on the `Rule` here

File: rust/parser/literal.rs [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+    debug_assert_eq!(node.as_rule(), Rule::quoted_string_literal);
+    StringLiteral { value: node.as_str().to_owned() }
+}
+
+fn visit_signed_integer(node: Node<'_>) -> SignedIntegerLiteral {
+    debug_assert_eq!(node.as_rule(), Rule::signed_integer);
+    let mut children = node.into_children().collect::<Vec<_>>();
+    let integral = children.pop().unwrap();
+    let sign = children.pop().map(|node| visit_sign(node));
+    debug_assert_eq!(integral.as_rule(), Rule::integer_literal);
```
Comment (dmitrii-ubskii) [inline rust/parser/literal.rs:59]: Take a look at other visitors, `children` can and should be used as an iterator.

File: rust/parser/pipeline.rs [hunk 15]
```diff
@@ -12,14 +12,15 @@ use super::{
...
-    visit_integer_literal, visit_label, visit_var, visit_vars, IntoChildNodes, Node, Rule, RuleMatcher,
+    visit_label, visit_var, visit_vars, IntoChildNodes, Node, Rule, RuleMatcher,
 };
 use crate::{
     common::{
         error::TypeQLError,
         token::{Aggregate, Order},
         Spanned,
     },
+    parser::literal::visit_integer_literal,
```
Comment (dmitrii-ubskii) [inline rust/parser/pipeline.rs:23]: `use super::literal::visit_integer_literal`

File: rust/value.rs [hunk 16]
```diff
@@ -4,12 +4,7 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
  */
 
-use std::{
-    fmt,
-    fmt::{Debug, Formatter},
-};
-use std::fmt::Write;
-use std::path::Display;
+use std::fmt::{self, Formatter};
```
Comment (dmitrii-ubskii) [inline rust/value.rs:7]:
```suggestion
use std::fmt;
```

File: rust/value.rs (hunk 10 above)
Comment (dmitrii-ubskii) [reply rust/value.rs:318]: This should probably now be called `unescape` or `unquote`

**Review by dmitrii-ubskii — CHANGES_REQUESTED** [review]

File: rust/parser/literal.rs [hunk 17]
```diff
@@ -107,12 +120,18 @@ fn visit_date_fragment(node: Node<'_>) -> DateFragment {
...
+        Some(children.consume_expected(Rule::second).as_str().to_owned())
+    } else {
+        None
+    };
+    let second_fraction = if children.peek().is_some() {
+        Some(children.consume_expected(Rule::second_fraction).as_str().to_owned())
+    } else {
+        None
+    };
     TimeFragment { hour, minute, second, second_fraction }
```
Comment (dmitrii-ubskii) [inline rust/parser/literal.rs:136]: One thing that's missing in these visitors is the assertion that the `children` iterator is fully exhausted. (`debug_assert_eq!(children.try_consume_any(), None)`)

File: rust/value.rs [hunk 18]
```diff
@@ -315,7 +315,7 @@ impl fmt::Display for StructLiteral {
 }
 
 impl StringLiteral {
-    pub fn parse_to_string(escaped_string: &str) -> Result<String> {
+    pub fn get_unescaped(escaped_string: &str) -> Result<String> {
```
Comment (dmitrii-ubskii) [inline rust/value.rs:318]: Just `unescape()`

**Review by dmitrii-ubskii — APPROVED** [review]
Couple more minor tweaks, otherwise LGTM

**Pull request merged**